kernel/ls3_fs.c

```diff
@@ -0,0 +1,1058 @@
+                    htonl(data[i+0]), htonl(data[i+1]), htonl(data[i+2]), htonl(data[i+3]),
+                    htonl(data[i+4]), htonl(data[i+5]), htonl(data[i+6]), htonl(data[i+7]),
+                    buf, 4*i);
+    }
+}
+
+static void dump_all_blocks_hex(int mountno, bool skip_empty) {
+    struct mount_data *m = mnt[mountno];
+    pr_info("Contents of device for mount[%d] (%llu blocks):\n",
+            mountno, m ? m->num_blocks : 0);
+    if (m == NULL) {
+        pr_info("  not yet mounted\n");
+        return;
+    }
+    void *blk = make_block();
+    uint64_t num_skipped = 0;
+    uint64_t i;
+    for (i = 0; i < m->num_blocks; i++) {
+        if (skip_empty) {
+            if (!test_bit(i, m->freemap)) {
+                num_skipped++;
+                continue;
+            }
+            if (num_skipped > 0) {
+                pr_info(" ... (skipped %lld unused block%s) ...\n",
+                        num_skipped, num_skipped > 1 ? "s" : "");
+            }
+            num_skipped = 0;
+        }
+        get_block(i, blk, mountno);
+        dump_block_hex(blk, i);
+    }
+    if (num_skipped > 0) {
+        pr_info(" ... (skipped %lld unused block%s) ...\n",
+                num_skipped, num_skipped > 1 ? "s" : "");
+    }
+}
+
+static void dump_bitmap(unsigned long *freemap, uint64_t num_bits) {
+    uint64_t i, j = 0;
+    char buf[33];
+    buf[32] = '\0';
+    int pos = 0;
+    for (i = 0; i < num_bits; i++) {
+        buf[pos++] = test_bit(i, freemap) ? '1' : '0';
+        if (pos == 32) {
+            pr_info("  %s status for blocks %llu to %llu\n", buf, j, i);
+            pos = 0;
+            j = i+1;
+        }
+    }
+    if (pos != 0) {
+        while (pos < 32)
+            buf[pos++] = ' ';
+        pr_info("  %s status for blocks %llu to %llu\n", buf, j, i-1);
+    }
+}
+
+// NEED PARAM, ARE WE RESERVING SUPER OR DATA OR FREE? data is 35 onward+
+// #define BLOCK_TYPE_SUPER    (1) // will hold super_block, in blocks 0 .. 31
+// #define BLOCK_TYPE_FREEMAP  (2) // will hold freemap data, in blocks 32 .. X
+// #define BLOCK_TYPE_KEYS     (3) // will hold key_block, in blocks X .. end
+// #define BLOCK_TYPE_DATA     (4) // will hold data_block, in blocks X .. end
+static uint64_t reserve_block(int mountno) {
+    // search the free freemap, find a 1, change 1 to 0 and return corresponding number that goes with the block.
+
+    uint64_t num_blocks = mnt[mountno]->num_blocks;
+    if (verbose >= 3) {
+        pr_info("Reserving one of %lld blocks, freemap is:\n", num_blocks);
+        dump_bitmap(mnt[mountno]->freemap, num_blocks);
+    }
+
+    int order = 0; // 2^order = 1, looking for region of size 1 bit
+    uint64_t blockno = bitmap_find_free_region(mnt[mountno]->freemap, num_blocks, order);
+
+    uint64_t freemap_array_idx = blockno / (8 * PAGE_CACHE_SIZE);
+    bitmap_set(mnt[mountno]->freemap_array_status, freemap_array_idx, 1);
```

Comment:
this is unnecessary, the above function sets the region when it finds it I think